servicetalk-http-api/src/main/java/io/servicetalk/http/api/StreamingHttpServiceToBlockingStreamingHttpService.java

```diff
@@ -80,102 +86,89 @@ public void closeGracefully() throws Exception {
+            private Subscription subscription;
+            private final int demandBatchSize;
+            private int outstandingDemand;
 
-            PayloadPump(final Subscriber subscriber, final HttpPayloadWriter<Buffer> payloadWriter) {
+            PayloadPump(final Subscriber subscriber, final HttpPayloadWriter<Buffer> payloadWriter,
+                        final int demandBatchSize) {
                 this.subscriber = subscriber;
                 this.payloadWriter = payloadWriter;
+                this.demandBatchSize = demandBatchSize;
             }
 
             @Override
-            public void onSubscribe(final PublisherSource.Subscription inSubscription) {
-                // We need to protect sub.cancel() from concurrent invocation with sub.request(MAX)
-                subscriber.onSubscribe(this);
-                inSubscription.request(Long.MAX_VALUE);
-                delayedCancellable(inSubscription);
+            public void onSubscribe(final Subscription inSubscription) {
+                // We need to protect sub.cancel() from concurrent invocation.
+                subscription = ConcurrentSubscription.wrap(inSubscription);
+                subscriber.onSubscribe(subscription);
+                outstandingDemand = demandBatchSize;
+                subscription.request(outstandingDemand);
             }
 
             @Override
             public void onNext(@Nullable final Object bufferOrTrailers) {
-                assert bufferOrTrailers != null;
-                try {
-                    if (bufferOrTrailers instanceof Buffer) {
-                        payloadWriter.write((Buffer) bufferOrTrailers);
-                        return;
-                    }
-                    if (bufferOrTrailers instanceof HttpHeaders) {
-                        payloadWriter.setTrailers((HttpHeaders) bufferOrTrailers);
-                        return;
-                    }
-                    assert false : "Expected only buffer or trailer in payloadBodyAndTrailers()";
-                } catch (IOException e) {
+                if (bufferOrTrailers instanceof Buffer) {
                     try {
-                        if (tryTerminate()) {
-                            subscriber.onError(e);
-                        } else {
-                            throwException(e);
-                        }
-                    } finally {
-                        this.cancel();
+                        payloadWriter.write((Buffer) bufferOrTrailers);
+                    } catch (IOException e) {
+                        throwException(e);
                     }
+                } else if (bufferOrTrailers instanceof HttpHeaders) {
+                    payloadWriter.setTrailers((HttpHeaders) bufferOrTrailers);
+                } else {
+                    throw new IllegalArgumentException("unsupported type: " + bufferOrTrailers);
+                }
+                if (--outstandingDemand < demandBatchSize >>> 1) {
```

Comment:
If `demandBatchSize == 1` the result of `demandBatchSize >>> 1` will be 0, leading to incorrect behavior.
This logic is similar to what we have in `PublisherAsBlockingIterable`. Consider making them consistent: use `itemsToNextRequest = max(1, demandBatchSize / 2)` and method `requestMoreIfRequired()`.

Reply:
good catch! I will make consistent.

Reply:
I noticed `PublisherAsBlockingIterable` does some strange stuff relative to back pressure and request N management. I updated it to use a consistent approach in both spots: https://github.com/apple/servicetalk/pull/1331